File: src/main/java/io/github/gravetii/validation/GameTimeValidator.java
package io.github.gravetii.validation;

import io.github.gravetii.model.GameTime;
import io.github.gravetii.util.Pair;
import io.github.gravetii.validation.ValidationRule;

public class GameTimeValidator {
  public GameTime validate(Pair<String, String> obj) {
    return this.gameTimeParser()
        .validate(
            this.integerParser()
                .validate(this.isNumberRule().validate(this.nonEmptyRule().validate(obj))));
  }

  private ValidationRule<Pair<String, String>, Pair<String, String>> nonEmptyRule() {
    return obj -> {
      if (obj == null) return null;
      String min = obj.getFirst().isEmpty() ? "0" : obj.getFirst();
      String sec = obj.getSecond().isEmpty() ? "0" : obj.getSecond();
      return new Pair<>(min, sec);
    };
  }

  private ValidationRule<Pair<String, String>, Pair<String, String>> isNumberRule() {
    return obj -> {
      if (obj == null) return null;
      String min = obj.getFirst();
      String sec = obj.getSecond();
      return (min.matches("[0-9]+") && sec.matches("[0-9]+")) ? obj : null;
    };
  }

  private ValidationRule<Pair<String, String>, Pair<Integer, Integer>> integerParser() {
    return obj -> {
      if (obj == null) return null;
      try {
        int minutes = Integer.parseInt(obj.getFirst());
        int seconds = Integer.parseInt(obj.getSecond());
        return new Pair<>(minutes, seconds);
      } catch (NumberFormatException e) {
        return null;
      }
    };
  }

  private ValidationRule<Pair<Integer, Integer>, GameTime> gameTimeParser() {
    return obj -> {
      if (obj == null) return null;
      int minutes = obj.getFirst();
      int seconds = obj.getSecond();
      if (minutes == 0 && seconds == 0) return null;
      if (seconds < 0 || seconds > 60) return null;
      return new GameTime(minutes, seconds);
    };
  }
}

File: src/main/kotlin/io/github/gravetii/validation/GameTimeValidator.kt
package io.github.gravetii.validation

import io.github.gravetii.model.GameTime
import kotlin.NumberFormatException

object GameTimeValidator {

    private fun nonEmpty(obj: Pair<String, String>?): Pair<String, String>? {
        return obj?.let {
            Pair(it.first.ifEmpty { "0" }, it.second.ifEmpty { "0" })
        }
    }

    private fun isNumber(obj: Pair<String, String>?): Pair<Int, Int>? {
        return obj?.let {
            val r = Regex("[0-9]+")
            return if (it.first.matches(r) && it.second.matches(r)) {
                try {
                    Pair(it.first.toInt(), it.second.toInt())
                } catch (e: NumberFormatException) {
                    null
                }
            } else null
        }
    }

    private fun gameTimeParser(obj: Pair<Int, Int>?): GameTime? {
        return obj?.let {
            return if (it.first == 0 && it.second == 0) null
            else if (it.first < 0 || it.second > 60) null
            else GameTime(it.first, it.second)
        }
    }

    fun validate(obj: Pair<String, String>?): GameTime? {
        return gameTimeParser(isNumber(nonEmpty(obj)))
    }

}